File: c/CRC32/crc32.c
#include "crc32.h"

// IEEE 802.3 CRC-32 Polynomial: 0xEDB88320 (Reflected/Reversed representation of 0x04C11DB7)
#define CRC32_POLY 0xEDB88320

static uint32_t crc32_table[256];
static int table_initialized = 0;

void crc32_init(void) {
    uint32_t crc;
    int i, j;

    if (table_initialized) return;

    for (i = 0; i < 256; i++) {
        crc = i;
        for (j = 0; j < 8; j++) {
            if (crc & 1) {
                crc = (crc >> 1) ^ CRC32_POLY;
            } else {
                crc = crc >> 1;
            }
        }
        crc32_table[i] = crc;
    }

    table_initialized = 1;
}

uint32_t crc32(uint32_t crc, const uint8_t *buf, size_t len) {
    if (!table_initialized) {
        crc32_init();
    }

    // 常见的 CRC32 实现中，初始值通常为 0xFFFFFFFF
    // 这里由调用者传入，以便支持分段计算。
    // 如果是第一次调用，调用者应传入 0xFFFFFFFF (或根据协议传入 0)
    // 注意：这里的算法是基于 Table 的 Reflected 实现
    
    // crc = ~crc; // 通常在外部处理或者由调用者传入 ~0

    while (len--) {
        crc = (crc >> 8) ^ crc32_table[(crc ^ *buf++) & 0xFF];
    }

    // return ~crc; // 同样，是否取反由调用者决定或者提供 helper
    return crc;
}

uint32_t crc32_total(const uint8_t *buf, size_t len) {
    uint32_t crc = 0xFFFFFFFF;
    crc = crc32(crc, buf, len);
    return crc ^ 0xFFFFFFFF;
}

File: js/CRC32/CRC32.js
// IEEE 802.3 CRC-32 Polynomial: 0xEDB88320 (Reflected)
const CRC32_POLY = 0xEDB88320;

let crc32_table = null;

function crc32_init() {
    if (crc32_table !== null) return;

    crc32_table = new Uint32Array(256);
    for (let i = 0; i < 256; i++) {
        let crc = i;
        for (let j = 0; j < 8; j++) {
            if (crc & 1) {
                crc = (crc >>> 1) ^ CRC32_POLY;
            } else {
                crc = crc >>> 1;
            }
        }
        crc32_table[i] = (crc >>> 0); // Ensure unsigned 32-bit
    }
}

/**
 * Calculate CRC32
 * @param {number} crc - Initial value (usually 0xFFFFFFFF or result from previous call)
 * @param {Uint8Array|string} buf - Data buffer or string
 * @returns {number} - The updated CRC register value (not inverted yet)
 */
function crc32(crc, buf) {
    if (crc32_table === null) {
        crc32_init();
    }

    if (typeof buf === 'string') {
        buf = new TextEncoder().encode(buf);
    }

    // Ensure crc is unsigned 32-bit
    crc = crc >>> 0;

    for (let i = 0; i < buf.length; i++) {
        const tableIndex = (crc ^ buf[i]) & 0xFF;
        crc = (crc >>> 8) ^ crc32_table[tableIndex];
    }

    return (crc >>> 0);
}

/**
 * Calculate total CRC32 (Standard implementation)
 * @param {Uint8Array|string} buf - Data buffer
 * @returns {number} - The final inverted CRC32 value
 */
function crc32_total(buf) {
    let crc = 0xFFFFFFFF;
    crc = crc32(crc, buf);
    return (crc ^ 0xFFFFFFFF) >>> 0;
}

module.exports = {
    crc32_init,
    crc32,
    crc32_total
};
